### PR TITLE
[RLlib] Request CPU resources in `Trainer.default_resource_request()` if using dataset input.

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

We need to request CPU resource when using dataset as offline data input.
Also clean up things a little bit.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
 from ray.rllib.execution.train_ops import TrainOneStep, MultiGPUTrainOneStep, \
     train_one_step, multi_gpu_train_one_step
 from ray.rllib.models import MODEL_DEFAULTS
+from ray.rllib.offline import get_offline_io_resource_bundles
```
Comment (sven1977) [inline rllib/agents/trainer.py:40]: Nit: Full path to imported module.

File: rllib/offline/resource.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from ray.rllib.offline.dataset_reader import get_resource_bundles \
+    as dataset_reader_get_resource_bundles
+from ray.rllib.utils.typing import PartialTrainerConfigDict
+from typing import Dict, List
+
+
+def get_offline_io_resource_bundles(config: PartialTrainerConfigDict) -> \
+        List[Dict]:
```
Comment (sven1977) [inline rllib/offline/resource.py:8]: Nit: ` -> List[Dict[str, float]]`?

File: rllib/offline/dataset_reader.py [hunk 3]
```diff
@@ -20,6 +77,9 @@ class DatasetReader(InputReader):
             "input_config"={
                 "format": "json",
                 "path": "/tmp/sample_batches/",
+                # By default, parallelism=num_workers.
```
Comment (sven1977) [inline rllib/offline/dataset_reader.py:80]: Same comment as the PR before (we should do this in some follow-up PR), just for documentation purposes:
```
input_config:
    type: "DatasetReader",
    [c'tor kwarg1]: ...,
    [c'tpr kwarg2]: ...,
```

File: rllib/offline/dataset_reader.py [hunk 4]
```diff
@@ -20,6 +77,9 @@ class DatasetReader(InputReader):
...
+                "parallelism": 3,
+                "num_cpus_per_read_task": 0.5,
```
Comment (sven1977) [inline rllib/offline/dataset_reader.py:82]: Just a question for my understanding: CPUs=0.5 means this is not fully parallelized, then, just like python-threading (of having two processes run on same CPU)?

**Review by sven1977 — APPROVED** [review]
Looks great, just a few nits and questions before we can merge.

File: rllib/agents/trainer.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/trainer.py:58]: sorry, I bit confused, this is the full path I thought?
can you clarify a bit? thanks.

File: rllib/offline/dataset_reader.py (hunk 3 above)
Comment (gjoliver) [reply rllib/offline/dataset_reader.py:83]: sorry, I must have missed something.
can you tell me which PR it is?
I'd love to fix documentation for sure.
I actually need to put up PR to add documentation for DatasetReader once this PR is in.

File: rllib/offline/dataset_reader.py (hunk 4 above)
Comment (gjoliver) [reply rllib/offline/dataset_reader.py:85]: oh, Dataset spins up a Ray task for each shard, reading all the data into memory.
this basically means 2 such shard readers can be scheduled on a single CPU.
0.5 is the default that Dataset uses.
I think for smaller RL dataset, we can even use 0.1, or 0.01 actually.

File: rllib/offline/resource.py (hunk 2 above)
Comment (gjoliver) [reply rllib/offline/resource.py]: done

**Review by gjoliver — COMMENTED** [review]
thanks for the review.

File: rllib/agents/trainer.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/trainer.py:58]: Sorry, I meant:
```
from ray.rllib.offline.resource import get_offline_io_resource_bundles
```

**Pull request merged**